stackslib/src/chainstate/nakamoto/mod.rs

```diff
@@ -0,0 +1,1375 @@
+        tenure_tx_fees: u128,
+    ) -> Result<(), ChainstateError> {
+        assert_eq!(tip_info.stacks_block_height, header.chain_length,);
+        assert!(tip_info.burn_header_timestamp < i64::MAX as u64);
+
+        let index_root = &tip_info.index_root;
+        let consensus_hash = &tip_info.consensus_hash;
+        let burn_header_hash = &tip_info.burn_header_hash;
+        let block_height = tip_info.stacks_block_height;
+        let burn_header_height = tip_info.burn_header_height;
+        let burn_header_timestamp = tip_info.burn_header_timestamp;
+
+        let block_size_str = format!("{}", tip_info.anchored_block_size);
+
+        let block_hash = header.block_hash();
+
+        let index_block_hash =
+            StacksBlockHeader::make_index_block_hash(&consensus_hash, &block_hash);
+
+        assert!(block_height < (i64::MAX as u64));
+
+        let args: &[&dyn ToSql] = &[
+            &u64_to_sql(block_height)?,
+            &index_root,
+            &consensus_hash,
+            &burn_header_hash,
+            &burn_header_height,
+            &u64_to_sql(burn_header_timestamp)?,
+            &block_size_str,
+            &HeaderTypeNames::Nakamoto,
+            &header.version,
+            &u64_to_sql(header.chain_length)?,
+            &u64_to_sql(header.burn_spent)?,
+            &header.parent,
+            &header.burn_view,
+            &header.signature,
+            &header.tx_merkle_root,
+            &header.state_index_root,
+            &block_hash,
+            &index_block_hash,
+            anchored_block_cost,
+            total_tenure_cost,
+            &tenure_tx_fees.to_string(),
+            parent_id,
+            &u64_to_sql(affirmation_weight)?,
+            &u64_to_sql(tenure_height)?,
+            if tenure_changed { &1i64 } else { &0 },
+        ];
+
+        tx.execute(
+            "INSERT INTO nakamoto_block_headers
+                    (block_height,  index_root, consensus_hash,
+                     burn_header_hash, burn_header_height,
+                     burn_header_timestamp, block_size,
+
+                     header_type,
+                     version, chain_length, burn_spent, parent,
+                     burn_view, signature, tx_merkle_root, state_index_root,
+
+                     block_hash,
+                     index_block_hash,
+                     cost,
+                     total_tenure_cost,
+                     tenure_tx_fees,
+                     parent_block_id,
+                     affirmation_weight,
+                     tenure_height,
+                     tenure_changed)
+                    VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9, ?10, ?11, ?12, ?13, ?14, ?15, ?16, ?17, ?18, ?19, ?20, ?21, ?22, ?23, ?24, ?25)",
+            args
+        )?;
+
+        Ok(())
+    }
+
+    /// Append a Stacks block to an existing Stacks block, and grant the miner the block reward.
+    /// Return the new Stacks header info.
+    pub fn advance_tip(
```

Comment:
There are a few functions here (like this one) that seem to have the same names/args as those in `StacksChainState`. Maybe we should define a `ChainState` trait?

Reply:
Perhaps, but can we wait to do this until we're certain about the interfaces to these two structs?  It might be that they look the same right now simply because the Nakamoto ones are derived from the 2.x ones (but may change substantially later).

Reply:
Yep, I agree -- I'd rather wait to define an interface until its clear something wants to consume it.